examples/module/ipddp/train_ipddp_invpend.py

```diff
@@ -0,0 +1,163 @@
+    S = torch.tile(torch.zeros(ns, nc, device=device), (n_batch, T, 1, 1))
+    c = torch.tile(torch.zeros(1, device=device), (n_batch, T))
+
+    # Create constraint object
+    gx = torch.tile(torch.zeros(2*nc, ns, device=device), (n_batch, T, 1, 1))
+    gu = torch.tile(torch.vstack((torch.eye(nc, nc, device=device), - torch.eye(nc, nc, device=device)) ), (n_batch, T, 1, 1))
+    g = torch.tile(torch.hstack((-0.25 * torch.ones(nc, device=device), -0.25 * torch.ones(nc, device=device)) ), (n_batch, T, 1))
+    batch_lincon = pp.module.LinCon(gx, gu, g)
+
+
+    expert = dict(
+        param = torch.Tensor([10.0]).to(device)
+    )
+    fname = os.path.join(args.save, 'expert.pkl')
+    with open(fname, 'wb') as fi:
+        pkl.dump(expert, fi)
+
+    torch.manual_seed(args.seed)
+    param = torch.Tensor([8.0]).to(device).requires_grad_()
+
+    state_all =      torch.zeros(n_batch, T+1,  ns, device=device)
+    input_all = 0.02*torch.ones(n_batch,    T,  nc, device=device)
+    state_all[...,0,:] = state
+
+    init_traj = {'state': state_all,
+                 'input': input_all}
+
+    fname = os.path.join(args.save, 'pypose losses.csv')
+    loss_f = open(fname, 'w')
+    loss_f.write('im_loss,mse\n')
+    loss_f.flush()
+
+    fname = os.path.join(args.save, 'pypose time.csv')
+    time_d = open(fname, 'w')
+    time_d.write('backward_time,overall_time\n')
+    time_d.flush()
+
+    def get_loss(_param):
+        sys_ = InvPend(dt, length=_param)
+        _ipddp_list = [None for batch_id in range(n_batch)]
+        _fp_best_list = [None for batch_id in range(n_batch)]
+        for batch_id in range(n_batch): # solved separated
+            stage_cost = pp.module.QuadCost(Q[batch_id:batch_id+1], R[batch_id:batch_id+1], S[batch_id:batch_id+1], c[batch_id:batch_id+1])
+            terminal_cost = pp.module.QuadCost(10./dt*Q[batch_id:batch_id+1,0:1,:,:], R[batch_id:batch_id+1,0:1,:,:], S[batch_id:batch_id+1,0:1,:,:], c[batch_id:batch_id+1,0:1])
+            lincon = pp.module.LinCon(gx[batch_id:batch_id+1], gu[batch_id:batch_id+1], g[batch_id:batch_id+1])
+            init_traj_sample = {'state': init_traj['state'][batch_id:batch_id+1],
+                                'input': init_traj['input'][batch_id:batch_id+1]}
+            _ipddp_list[batch_id] = IPDDP(sys_, stage_cost, terminal_cost, lincon,
+                                    gx.shape[-2], init_traj_sample)
+            # detached version to solve the best traj
+            with torch.no_grad():
+                _fp_best_list[batch_id] = _ipddp_list[batch_id].solver()
+
+        x_pred, u_pred, _ = _ipddp_list[0].forward(_fp_best_list) # call any one class instantiation, but perform batch grad computation
+        print('x_pred solved')
+        sys = InvPend(dt, length=expert['param'])
+        ipddp_list = [None for batch_id in range(n_batch)]
+        fp_list = [None for batch_id in range(n_batch)]
+        for batch_id in range(n_batch):
```

Comment:
There should be some better way to handle the batch scenario. I suggest we focus on the non-batched case first.

Reply:
Yes, currently the solver() can only solve a single trajectory, but the autodiff part using forward() supports batched case; so I leave the interface of the batched case here. 
If the users are interested in the non-batched case, they can just set the state with n_batch = 1.